changelog/deprecate_bad_module_name.dd

```diff
@@ -0,0 +1,47 @@
+Deprecate import module name mismatch
+
+Some instances of module name/import name mismatches were deprecated while others are now considered errors.
+
+Case 1: DEPRECATED: import a module name with a qualified name but the module being imported has no module declaration.
+---
+// main.d
+import foo.bar;
+
+// foo/bar.d
+// this file is empty, it has no module declaration
+---
+The above code will now print:
+
+$(CONSOLE Deprecated: module bar from file foo/bar.d should be imported with 'import bar;' )
+
+Case 2: DEPRECATED: import a module with a qualified name that partially matches the name of the module being imported.
+---
+// main.d
+import foo.bar;
+import foo.baz.buz;
+
+// foo/bar.d
+module bar;
+
+// foo/baz/buz.d
+module baz.buz;
+---
+The above code will now print:
+
+$(CONSOLE Deprecated: module bar from file foo/bar.d should be imported with 'import bar;'
+Deprecated: module baz.buz from file foo/baz/buz.d should be imported with 'import baz.buz;' )
+
+Note that for this rule to apply, the shorter name must completely match the end of the longer name.
+
+Case 3: ERROR: import a module that matches the filename but does not match the module name.
+---
+// main.d
+
+// foo.d
+module bar;
+---
+The above code will now fail and print:
+
+$(CONSOLE Error: module bar from file foo.d must be imported with 'import bar;')
```

Comment:
I don't understand how the file would ever be found. So the compiler is supposed to compile `main.d`. It sees `import bar;`. How does it know it's supposed to look for a file called `foo.d`?

Reply:
The file is give on the command line.  Module names not matching file names is a technique already used, though the import statement will match the module name (just no the filename).  You have 3 different names:

1. Filename
2. Import Name
3. Module Name

This PR enforces that "import name" and "module name" should match, but filename does not necessarily need to.  @jacob-carlborg provided an example where this is useful (https://github.com/dlang/dmd/pull/7778#issuecomment-360706239)